oci/oci_dest.go

```diff
@@ -0,0 +1,135 @@
+package oci
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"regexp"
+	"strings"
+
+	"github.com/projectatomic/skopeo/manifest"
+	"github.com/projectatomic/skopeo/reference"
+	"github.com/projectatomic/skopeo/types"
+)
+
+type descriptor struct {
+	Digest    string `json:"digest"`
+	MediaType string `json:"mediaType"`
+	Size      int64  `json:"size"`
+}
+
+type ociImageDestination struct {
+	dir string
+	ref reference.Named
+	tag string
+}
+
+var refRegexp = regexp.MustCompile(`^([A-Za-z0-9._-]+)+$`)
+
+// NewOCIImageDestination returns an ImageDestination for writing to an existing directory.
+func NewOCIImageDestination(dir string) (types.ImageDestination, error) {
+	dest := dir
+	sep := strings.LastIndex(dir, ":")
+	tag := "latest"
+	if sep != -1 {
+		dest = dir[:sep]
+		tag = dir[sep+1:]
+		if !refRegexp.MatchString(tag) {
+			return nil, fmt.Errorf("Invalid reference %s", tag)
+		}
+	}
+	return &ociImageDestination{
+		dir: dest,
+		tag: tag,
+	}, nil
+}
+
+func (d *ociImageDestination) CanonicalDockerReference() (string, error) {
+	return "", fmt.Errorf("Can not determine canonical Docker reference for an OCI image")
+}
+
+func (d *ociImageDestination) PutManifest(m []byte) error {
+	if err := d.ensureParentDirectoryExists("refs"); err != nil {
+		return err
+	}
+	// TODO(runcom):
+	// make a function to create an OCI manifest from a distribution one which comes here!!!
+	// convert the manifest to OCI v1 - beaware we can also receive a docker v2s1 manifest so we need to
+	// handle this! docker.io/library/busybox it's still on v2s1 for instance
+	digest, err := manifest.Digest(m)
+	if err != nil {
+		return err
+	}
+	desc := descriptor{}
+	desc.Digest = digest
+	mt := manifest.GuessMIMEType(m)
+	// TODO(runcom): beaware and add support for OCI manifest list
```

Comment:
I haven’t thought about manifests that much. I am leaning towards treating them as a repository of independent images rather than as a single image with multiple sets of data (what does it mean to “read the $name annotation” if there is a manifest list and multiple annotated manifests?), but I’m not sure that is intuitive to users either.